File: flowjet_valves/public/py/sales_order.py
import frappe
import json
from flowjet_valves.public.py.production_plan import get_items_for_material_requests
from erpnext.stock.doctype.item.item import get_item_defaults

@frappe.whitelist()
def update_manufacture_cycle(name, custom_priority=None, item_priority_map=None):
    if isinstance(item_priority_map, str):
        item_priority_map = json.loads(item_priority_map)

    # Step 1: Get related Production Plans from Sales Order
    production_plans = frappe.get_all(
        "Production Plan Sales Order",
        filters={"sales_order": name},
        fields=["parent"]
    )

    for plan in production_plans:
        production_plan = frappe.get_doc("Production Plan", plan.parent)
        
        # Skip cancelled Production Plans
        if production_plan.docstatus == 2:
            continue

        # Step 2: Update priority in Production Plan Sales Order table
        for row in production_plan.sales_orders:
            if row.sales_order == name:
                row.custom_priority = custom_priority

        # Step 3: Update custom_priority in Production Plan Item table
        for item in production_plan.po_items:
            if item.item_code in item_priority_map:
                item.custom_priority = item_priority_map[item.item_code]

                work_orders = frappe.get_list("Work Order", filters={"production_plan_item": item.name}, pluck="name")
                for wo_name in work_orders:
                    work_order = frappe.get_doc("Work Order", wo_name)
                    work_order.custom_priority = item.custom_priority
                    work_order.save(ignore_permissions=True)

                    job_cards = frappe.get_list("Job Card", filters={"work_order": wo_name}, pluck="name")
                    for jc_name in job_cards:
                        frappe.db.set_value("Job Card", jc_name, "custom_priority", item.custom_priority)


        # Step 4: Update custom_priority in Production Plan Sub Assembly Item table
        for item in production_plan.sub_assembly_items:
            if item.parent_item_code in item_priority_map:

                work_order_sub_assemblies = frappe.get_list("Work Order", filters={"production_plan_sub_assembly_item": item.name}, pluck="name")
                for wo_name in work_order_sub_assemblies:
                    work_order = frappe.get_doc("Work Order", wo_name)
                    work_order.custom_priority = item_priority_map[item.parent_item_code]
                    work_order.save(ignore_permissions=True)

                    job_cards = frappe.get_list("Job Card", filters={"work_order": wo_name}, pluck="name")
                    for jc_name in job_cards:
                        frappe.db.set_value("Job Card", jc_name, "custom_priority", item_priority_map[item.parent_item_code])

        production_plan.save(ignore_permissions=True)

        # Step 5: Update Material Request priorities via MR item generation
        items = get_items_for_material_requests(production_plan.as_dict())

        for item in items:
            item_code = item.get("item_code")
            priority = item.get("custom_priority")

            # Find matching Material Request Plan Item rows (you can narrow down further using production_plan or company if needed)
            mr_items = frappe.get_all(
                "Material Request Plan Item",
                filters={
                    "item_code": item_code,
                    "parent": production_plan.name
                },
                fields=["name"]
            )

            for mr_item in mr_items:
                frappe.db.set_value("Material Request Plan Item", mr_item.name, "custom_priority", priority)


@frappe.whitelist()
def get_sales_order_items(sales_order):
    items = frappe.get_all(
        "Sales Order Item",
        filters={"parent": sales_order},
        fields=["name", "item_code", "item_name", "qty", "warehouse", "delivery_date", "actual_qty"]
    )

    # filtered_items = []

    # for item in items:
    #     submitted_dn = frappe.db.sql("""
    #         SELECT SUM(qty) FROM `tabDelivery Note Item`
    #         WHERE sales_order_item = %s AND docstatus = 1
    #     """, item.name)[0][0] or 0

    #     remaining_qty = item.qty - submitted_dn

    #     # Set the submitted PO qty into custom_pending_for_po_qty
    #     frappe.db.set_value("Sales Order Item", item.name, "custom_pending_for_po_qty", remaining_qty)

    #     if remaining_qty <= 0:
    #         continue

    #     item["qty"] = remaining_qty

    #     item_defaults = get_item_defaults(
    #         item.item_code,
    #         frappe.db.get_value("Sales Order", sales_order, "company")
    #     )
    #     item["supplier"] = item_defaults.get("default_supplier")

    #     filtered_items.append(item)

    # return filtered_items
    return items

File: flowjet_valves/public/js/sales_order.js
frappe.ui.form.on('Sales Order', {
    custom_priority(frm) {
        if (frm.doc.items && frm.doc.custom_priority) {
            // updateProductionPlans(frm);
            frm.doc.items.forEach(row => {
                if (!row.custom_priority) row.custom_priority = frm.doc.custom_priority;
            });
            frm.refresh_field('items');
        }
    },
    after_save(frm) {
        if (frm.doc.docstatus == 1) updateManufactureCycle(frm);
    },
    refresh(frm) {
        if (!frm.doc.selling_price_list) return;

        frm.doc.items.forEach(row => {
            if (!row.item_code) return;

            frappe.call({
                method: "frappe.client.get_value",
                args: {
                    doctype: "Item Price",
                    filters: {
                        item_code: row.item_code,
                        price_list: frm.doc.selling_price_list
                    },
                    fieldname: "price_list_rate"
                },
                callback: function (r) {
                    if (!r.message) return;

                    let price_list_rate = r.message.price_list_rate;
                    if (flt(row.rate) !== flt(price_list_rate)) {
                        $(`[data-idx="${row.idx}"] [data-fieldname="rate"]`).css("color", "red");
                    } else {
                        $(`[data-idx="${row.idx}"] [data-fieldname="rate"]`).css("color", "");
                    }
                }
            });
        });
        if (frm.doc.docstatus == 1) frm.add_custom_button(__('Dispatch Allocation'), function() {
            frappe.call({
                method: "flowjet_valves.public.py.sales_order.get_sales_order_items",
                args: { sales_order: frm.doc.name },
                callback: function (r) {
                    if (!r.message || r.message.length === 0) {
                        frappe.msgprint("No items found in the Sales Order.");
                        return;
                    }

                    let items = r.message.map((item) => ({
                        item_code: item.item_code,
                        item_name: item.item_name,
                        qty: item.qty,
                        name: item.name,
                        warehouse: item.warehouse,
                        // batch_no: item.batch_no,
                        delivery_date: item.delivery_date,
                        actual_qty: item.actual_qty
                    }));
                    console.log("Items:", items);

                    let dialog = new frappe.ui.Dialog({
                        title: __("Create Dispatch Allocation"),
                        fields: [
                            {
                                fieldtype: "Data",
                                fieldname: "selected_count",
                                label: "Selected Items Count",
                                default: 0,
                                read_only: 1,
                            },
                            {
                                fieldtype: "Table",
                                fieldname: "item_table",
                                label: "Items",
                                cannot_add_rows: true,
                                cannot_delete_all_rows: true,
                                fields: [
                                    {
                                        fieldtype: "Data",
                                        fieldname: "item_code",
                                        label: "Item Code",
                                        read_only: 1,
                                        in_list_view: 1,
                                        columns: 2,
                                    },
                                    {
                                        fieldtype: "Float",
                                        fieldname: "qty",
                                        label: "Required Qty",
                                        read_only: 1,
                                        in_list_view: 1,
                                        columns: 1,
                                    },
                                    {
                                        fieldtype: "Float",
                                        fieldname: "allocated_qty",
                                        label: "Allocated Qty",
                                        read_only: 0,
                                        in_list_view: 1,
                                        columns: 1,
                                    },
                                    {
                                        fieldtype: "Float",
                                        fieldname: "actual_qty",
                                        label: "Available Qty",
                                        read_only: 1,
                                        in_list_view: 1,
                                        columns: 1,
                                    },
                                    {
                                        fieldtype: "Link",
                                        fieldname: "warehouse",
                                        label: "Warehouse",
                                        options: "Warehouse",
                                        read_only: 1,
                                        in_list_view: 1,
                                        columns: 2,
                                    },
                                    {
                                        fieldtype: "Link",
                                        fieldname: "batch_no",
                                        label: "Batch No.",
                                        options: "Batch",
                                        read_only: 1,
                                        in_list_view: 1,
                                        columns: 1,
                                    },
                                    {
                                        fieldtype: "Date",
                                        fieldname: "delivery_date",
                                        label: "Expected Dispatch Date",
                                        read_only: 1,
                                        in_list_view: 1,
                                        columns: 1,
                                    },
                                    {
                                        fieldtype: "Data",
                                        fieldname: "item_name",
                                        label: "Item Name",
                                        read_only: 1,
                                        columns: 3,
                                    },
                                ],
                            },
                        ],
                        primary_action_label: __("Create"),
                        primary_action(values) {
                        let selected_items = dialog.fields_dict.item_table.grid.get_selected_children();
                        if (!selected_items.length) {
                            frappe.msgprint("Please select at least one item.");
                            return;
                        }

                            dialog.hide();
                            frappe.model.open_mapped_doc({
                                method: "erpnext.selling.doctype.sales_order.sales_order.make_delivery_note",
                                frm: frm,
                                args: {
                                    // supplier: values.supplier, // pass selected supplier
                                    filtered_children: selected_items.map((row) => row.name), // pass only selected items
                                },
                                run_link_triggers: true,
                            });
                        },
                        // Add a custom class
                        wrapper_class: "wide-dispatch-dialog"
                    });

                    // Fill and show dialog
                    dialog.fields_dict.item_table.df.data = items;
                    dialog.fields_dict.item_table.grid.grid_pagination.page_length = 10;
                    dialog.fields_dict.item_table.grid.refresh();
                    dialog.show();
                    dialog.$wrapper.find(".modal-content").css("width", "250%");
                    dialog.$wrapper.find(".modal-dialog").css("margin", "1.75rem 17.5rem");

                // Count updater function
                function updateSelectedCount() {
                    let count = dialog.fields_dict.item_table.grid.get_selected().length;
                    dialog.set_value("selected_count", count);
                    dialog.refresh_field("selected_count");
                }

                // Bind event listener to track selection changes
                dialog.fields_dict.item_table.grid.wrapper.on('click', '.grid-row-check', function() {
                    setTimeout(() => {
                        updateSelectedCount();
                    }, 50);
                });

                // Initial update count
                setTimeout(() => {
                    updateSelectedCount();
                }, 100);
                }
            });
        }, __('Make'));
    },
});


frappe.ui.form.on('Sales Order Item', {
    rate(frm, cdt, cdn) {
        let row = locals[cdt][cdn];
        if (!row.item_code || !frm.doc.selling_price_list) return;

        frappe.call({
            method: "frappe.client.get_value",
            args: {
                doctype: "Item Price",
                filters: {
                    item_code: row.item_code,
                    price_list: frm.doc.selling_price_list
                },
                fieldname: "price_list_rate"
            },
            callback: function (r) {
                if (!r.message) return;

                let price_list_rate = r.message.price_list_rate;
                // Compare the rate
                if (price_list_rate && (flt(row.rate) !== flt(price_list_rate))) {
                    // Apply red style to the 'rate' field
                    $(`[data-idx="${row.idx}"] [data-fieldname="rate"]`).css("color", "red");
                } else {
                    // Reset color if it's the same
                    $(`[data-idx="${row.idx}"] [data-fieldname="rate"]`).css("color", "");
                }
            }
        });
    }
});


function updateManufactureCycle(frm) {
    let item_priority_map = {};

    (frm.doc.items || []).forEach(item => {
        if (item.item_code) {
            item_priority_map[item.item_code] = item.custom_priority;
        }
    });

    frm.call({
        method: 'flowjet_valves.public.py.sales_order.update_manufacture_cycle',
        args: {
            name: frm.doc.name,
            custom_priority: frm.doc.custom_priority,
            item_priority_map: item_priority_map
        }
    });
}